contributors/design-proposals/vertical-pod-autoscaler.md

```diff
@@ -0,0 +1,486 @@
+OOMs across multiple Pods at random time or (3) save resources over long periods
+of time.
+
+Apart from its own policy on how often a Pod can be evicted, the Updater also
+respects the Pod disruption budget, by using Eviction API to evict pods.
+
+The Updater only touches Pods that (1) point to a VPA with updatePolicy.mode set
+to “auto” and (2) belong to a Replica Set.
+
+Note: the Updater may in future be extracted to a separate process.
+
+### Recommendation model ###
+
+VPA controls the request (memory and CPU) of containers. In MVP it always sets
+the limit to infinity. It is not yet clear whether there is a use-case for VPA
+setting the limit.
+
+The request is calculated based on analysis of the current and revious runs of
+the container and other containers with similar properties (name, image,
+command, args).
+The recommendation model (MVP) assumes that the memory and CPU consumption are
+independent random variables with distribution equal to the one observed in the
+last N days (recommended value is N=8 to capture weekly peaks).
+A more advanced model in future could attempt to detect trends, periodicity and
+other time-related patterns.
+
+For CPU the objective is to **keep the fraction of time when the container usage
+exceeds a high percentage (e.g. 95%) of request below a certain threshold**
+(e.g. 1% of time).
+In this model the “CPU usage” is defined as mean usage measured over a short
+interval. The shorter the measurement interval, the better the quality of
+recommendations for spiky, latency sensitive workloads. Minimum reasonable
+resolution is 1/min, recommended is 1/sec.
+
+For memory the objective is to **keep the probability of the container usage
+exceeding the request in a specific time window below a certain threshold**
+(e.g. below 1% in 24h). The window must be long (≥ 24h) to ensure that evictions
+caused by OOM do not visibly affect (a) availability of serving applications
+(b) progress of batch computations (a more advanced model could allow user to
+specify SLO to control this).
+
+#### Handling OOMs ####
+When a container is evicted due to exceeding available memory, its actual memory
+requirements are not known (the amount consumed obviously gives the lower
+bound). This is modelled by translating OOM events to artificial memory usage
+samples by applying a "safety margin" multiplier to the last observed usage.
+
+### History Storage ###
+VPA defines data access API for providers of historical events and resource
+utilization. Initially we will use InfluxDB as the reference implementation of
+this API. Users will be able to plug their own implementations.
+
+History Storage is populated with real time updates of resources utilization and
+events from Heapster, similarly to the Recommender. The storage keeps at least
+8 days of data. This data is only used to initialize the Recommender on startup.
+
+### Open questions ###
+1. How to resolve conflicts if multiple VPA objects match a Pod.
```

Comment:
Pod metadata has the ownerReferences list (https://kubernetes.io/docs/api-reference/v1/definitions/#_v1_ownerreference). A references on this list can be marked as "controller" with a promise that there will be no more than one controller.
However we cannot directly reuse this mechanism as is. Presumably we should allow different agents to control different aspects of the pod - we would have to extend this mechanism to allow different types of controllers. Not sure if this is the right way.
Of course this still wouldn't prevent the user from creating a pod that is matched by 2 VPAs. It would merely allow us to choose one of the two as the "active" one for the given pod.